pkg/controllers/job/plugins/mpi/mpi.go

```diff
@@ -0,0 +1,168 @@
+/*
+Copyright 2019 The Volcano Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mpi
+
+import (
+	"flag"
+	"fmt"
+	"strings"
+
+	"github.com/golang/glog"
+
+	"k8s.io/api/core/v1"
+
+	vkv1 "volcano.sh/volcano/pkg/apis/batch/v1alpha1"
+	"volcano.sh/volcano/pkg/apis/helpers"
+	plugininterface "volcano.sh/volcano/pkg/controllers/job/plugins/interface"
+	"volcano.sh/volcano/pkg/controllers/job/plugins/ssh"
+	"volcano.sh/volcano/pkg/controllers/job/plugins/svc"
+)
+
+type mpi struct {
+	// Arguments given for the plugin
+	pluginArguments []string
+
+	clientset plugininterface.PluginClientset
+
+	ssh plugininterface.PluginInterface
+
+	// mpi launcher index, by default it is 0.
+	launcher int
```

Comment:
why do we need this? we already support multiple pod template.

Reply:
Need this to only set `MPI_HOST` env and mount hostfile to launcher pod

Reply:
we need to provide info to all pods.

Reply:
similar to svc.

Reply:
I donot want to add useless info to mpi workers, unless they are needed

Reply:
refer to `svc` plugin for that.

Reply:
IIUC, mpi worker will never need to connect to the launcher.  A plugin should only add useful info to the container.

Reply:
please do not have specific setting to mpi.

Reply:
we should not distinguish master/worker here.

Reply:
FYI: https://github.com/kubeflow/mpi-operator/blob/815f72098264e1c53743e2cdde3675b207247873/pkg/controllers/v1alpha2/mpi_job_controller.go#L1062-L1149 What mpi operator does is also not mount any host info in the worker.

We should distinguish launcher and worker, as we have to support slots per node as well. IMO, the launcher is the master, it will send commands to workers. But workers will not communicate to the launcher initiative.